### PR TITLE
Updated openmrs 3.x

### PR DESCRIPTION
**Pull request by mseaton**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/appointments/service/AppointmentsService.java [hunk 1]
```diff
@@ -30,6 +30,18 @@ public interface AppointmentsService {
     @Authorized({VIEW_APPOINTMENTS, MANAGE_APPOINTMENTS})
     List<Appointment> getAllAppointments(Date forDate);
 
+    @Transactional
+    @Authorized({VIEW_APPOINTMENTS, MANAGE_APPOINTMENTS})
+    List<Appointment> getAllCameEarlyAppointments(Date forDate);
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/appointments/service/AppointmentsService.java:35]: Why do you need MANAGE_APPOINTMENTS to get this list/data?

File: api/src/main/java/org/openmrs/module/appointments/service/AppointmentsService.java [hunk 2]
```diff
@@ -30,6 +30,18 @@ public interface AppointmentsService {
...
+
+    @Transactional
+    @Authorized({VIEW_APPOINTMENTS, MANAGE_APPOINTMENTS})
+    List<Appointment> getCompletedAppointments(Date forDate);
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/appointments/service/AppointmentsService.java:39]: Why do you need MANAGE_APPOINTMENTS to get this list/data?

File: api/src/main/java/org/openmrs/module/appointments/service/AppointmentsService.java [hunk 3]
```diff
@@ -30,6 +30,18 @@ public interface AppointmentsService {
...
+
+
+    @Transactional
+    @Authorized({VIEW_APPOINTMENTS, MANAGE_APPOINTMENTS})
+    List<Appointment> getAllAppointments(Date forDate, String status);
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/appointments/service/AppointmentsService.java:44]: Why do you need MANAGE_APPOINTMENTS to get this list/data?

File: api/src/main/java/org/openmrs/module/appointments/service/impl/AppointmentsServiceImpl.java [hunk 4]
```diff
@@ -189,6 +190,36 @@ public List<Appointment> getAllAppointments(Date forDate) {
...
+    @Override
+    public List<Appointment> getCompletedAppointments(Date forDate){
+        List<Appointment> appointments = appointmentDao.getCompletedAppointments(forDate);
+        return appointments.stream().filter(appointment -> !isServiceOrServiceTypeVoided(appointment)).collect(Collectors.toList());
+    }
+
+    @Transactional
+    @Override
+    public List<Appointment> getAllAppointments(Date forDate, String status) {
+        List<Appointment> appointments = new ArrayList<Appointment>();
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/appointments/service/impl/AppointmentsServiceImpl.java:210]: I don't love the status.trim check on these strings.  Can we look at a better way?

File: api/src/main/java/org/openmrs/module/appointments/service/impl/AppointmentsServiceImpl.java [hunk 5]
```diff
@@ -373,12 +404,12 @@ public Appointment reschedule(String originalAppointmentUuid, Appointment newApp
 
         try {
             //cancel the previous appointment
-            changeStatus(prevAppointment, AppointmentStatus.Cancelled.toString(), new Date());
+            changeStatus(prevAppointment, AppointmentStatus.Rescheduled.toString(), new Date());
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/appointments/service/impl/AppointmentsServiceImpl.java:407]: Defer to others on this, or whether Cancelled -> Scheduled is the desired approach, and determining whether an appointment is "rescheduled" is done by inspecting the audit history.

File: api/src/main/java/org/openmrs/module/appointments/util/DateUtil.java [hunk 6]
```diff
@@ -73,6 +73,16 @@ public static Date getEndOfDay() {
         calendar.set(Calendar.MILLISECOND, calendar.getMaximum(Calendar.MILLISECOND));
         return calendar.getTime();
     }
+
+    public static Date getStartOfDayByDate( Date date) {
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTime(date);
+        calendar.set(Calendar.HOUR_OF_DAY, calendar.getMinimum(Calendar.HOUR_OF_DAY));
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/appointments/util/DateUtil.java:80]: I don't know about using calendar.getMinimum here.  I'd think there must be a good, existing method to do this either in core, or in apache commons, or whereever.  Maybe use `DateUtils.truncate` from apache commons...

File: api/src/main/java/org/openmrs/module/appointments/util/PatientUtil.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.Objects;
+
+import org.openmrs.Patient;
+import org.openmrs.PersonAttribute;
+
+public class PatientUtil {
+
+    private static final String IDENTIFIER_NAME = "identifierName";
+    private static final String IDENTIFIER = "identifier";
+    private static final String TELEPHONE_CONTACT = "Telephone contact";
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/appointments/util/PatientUtil.java:16]: Not really sure where this is used, but "Telephone contact" is not suitably generic and not found in all implementations by this name.  We should find a better way to make this match the necessary use cases generically.

File: api/src/main/resources/liquibase.xml [hunk 8]
```diff
@@ -567,6 +567,33 @@
         </sql>
 
     </changeSet>
+    <changeSet id="add_visit_date_patient_appointment_20230222" author="makombe">
```
Comment (mseaton) [inline api/src/main/resources/liquibase.xml:570]: If we are merging this in from a PR, hopefully we can just merge this down into one changeSet that adds the right column with the right name.

File: omod/src/main/java/org/openmrs/module/appointments/web/contract/AppointmentsSummary.java [hunk 9]
```diff
@@ -4,17 +4,37 @@
...
     @JsonCreator
     public AppointmentsSummary(@JsonProperty("appointmentService")AppointmentServiceDefaultResponse appointmentService,
                                @JsonProperty("appointmentCountMap") Map appointmentCountMap) {
         this.appointmentService = appointmentService;
         this.appointmentCountMap = appointmentCountMap;
     }
+    @JsonCreator
+    public AppointmentsSummary(@JsonProperty("services") List services,
+                               @JsonProperty("appointmentDate") String appointmentDate,
+                               @JsonProperty("dailAappointmentCount") Integer totalServiceCount) {
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/appointments/web/contract/AppointmentsSummary.java:32]: this JsonProperty is spelled incorrectly

File: omod/src/main/java/org/openmrs/module/appointments/web/controller/AppointmentController.java [hunk 10]
```diff
@@ -50,91 +62,112 @@ public class AppointmentController extends BaseRestController {
 
     @RequestMapping(method = RequestMethod.GET, value = "all")
     @ResponseBody
-    public List<AppointmentDefaultResponse> getAllAppointments(@RequestParam(value = "forDate", required = false) String forDate) throws ParseException {
-        List<Appointment> appointments = appointmentsService.getAllAppointments(DateUtil.convertToLocalDateFromUTC(forDate));
+    public List<AppointmentDefaultResponse> getAllAppointments(
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/appointments/web/controller/AppointmentController.java:65]: All of the formatting changes here are hard to read through and know what are meaningful changes.  Can these formatting changes be removed or moved to a separate PR?

File: omod/src/main/java/org/openmrs/module/appointments/web/controller/AppointmentController.java [hunk 11]
```diff
@@ -147,48 +180,261 @@ public ResponseEntity<Object> undoStatusChange(@PathVariable("appointmentUuid")S
...
+            Appointment rescheduledAppointment = appointmentsService.reschedule(prevAppointmentUuid, appointment,
+                    false);
             return new ResponseEntity<>(appointmentMapper.constructResponse(rescheduledAppointment), HttpStatus.OK);
         } catch (RuntimeException e) {
             log.error("Runtime error while trying to create new appointment", e);
             return new ResponseEntity<>(RestUtil.wrapErrorResponse(e, e.getMessage()), HttpStatus.BAD_REQUEST);
         }
     }
 
+    /**
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/appointments/web/controller/AppointmentController.java:231]: Not reviewing these methods in detail at this stage.

File: omod/src/main/java/org/openmrs/module/appointments/web/mapper/AppointmentMapper.java [hunk 12]
```diff
@@ -272,9 +274,24 @@ private Map createPatientMap(Patient p) {
         map.put("name", p.getPersonName().getFullName());
         map.put("uuid", p.getUuid());
         map.put("identifier", p.getPatientIdentifier().getIdentifier());
-        map.put("age", p.getAge());
-        map.put("gender", p.getGender());
         map.putAll(p.getActiveIdentifiers().stream().filter(e -> e.getIdentifierType() != null).collect(Collectors.toMap(e -> e.getIdentifierType().toString().replaceAll("[- ]", ""), e -> e.getIdentifier())));
+        for (PatientIdentifier patientIdentifier : p.getIdentifiers()) {
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/appointments/web/mapper/AppointmentMapper.java:278]: This doesn't look right/appropriate to me.  Too much hard-coding.  Why don't we just use standard OpenMRS REST responses and conversions for this data?

File: omod/src/main/resources/config.xml [hunk 13]
```diff
@@ -60,21 +64,6 @@
 	</messages>
 	<!-- /Internationalization -->
 
-	<advice>
```
Comment (mseaton) [inline omod/src/main/resources/config.xml:63]: I suspect these cannot/should not be removed, and I'd hope that they are invoked optionally if one does not have the atom feed module installed then they have no impact.

File: omod/src/main/resources/webModuleApplicationContext.xml [hunk 14]
```diff
@@ -22,6 +22,10 @@
 
 	<!-- Add here beans related to the web context -->
 
-	<context:component-scan base-package="org.openmrs.module.appointments.web" />
+	<bean class="org.openmrs.ui.framework.StandardModuleUiConfiguration">
```
Comment (mseaton) [inline omod/src/main/resources/webModuleApplicationContext.xml:25]: Are we supporting the UI 2.x framework in this module?  Why is this being added?

File: api/src/main/java/org/openmrs/module/appointments/util/DateUtil.java (hunk 6 above)
Comment (ibacher) [reply api/src/main/java/org/openmrs/module/appointments/util/DateUtil.java:80]: Just use `java.time` for this:

```java
LocalDate.now().atStartOfDay();
```

If we really need this as a date:

```java
Date.from(LocalDate.now().atStartOfDay().atZone(ZoneId.systemDefault()).toInstant());
```

File: api/src/main/resources/Appointment.hbm.xml [hunk 15]
```diff
@@ -50,6 +50,7 @@
         <many-to-one name="creator" class="org.openmrs.User" column="creator" not-null="true"/>
         <many-to-one name="relatedAppointment" class="org.openmrs.module.appointments.model.Appointment"
                     column="related_appointment_id" unique="true"/>
+        <property name="dateHonored" type="java.util.Date" column="date_honored"/>
```
Comment (ibacher) [inline api/src/main/resources/Appointment.hbm.xml:53]: I'm a little hesitant about this. It seems like it would be better to have a link from a Visit -> an Appointment (which could be done via a Visit attribute) rather than tracking state that changes the lifetime of the appointment. This is less likely to lead to inconsistent data (where, for example, the appointments table indicates _X_ # of appointments were fulfilled, but there are _Y_ visits recorded in the system).

File: omod/src/main/java/org/openmrs/module/appointments/web/contract/AppointmentsSummary.java [hunk 16]
```diff
@@ -30,4 +50,29 @@ public Map<String, DailyAppointmentServiceSummary> getAppointmentCountMap() {
...
+
+    public String getAppointmentDate() {
+        return appointmentDate;
+    }
+
+    public void setAppointmentDate(String appointmentDate) {
+        this.appointmentDate = appointmentDate;
+    }
+
+    public Integer gettotalServiceCount() {
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/appointments/web/contract/AppointmentsSummary.java:70]: This applies to the setter as well:

```suggestion
    public Integer getTotalServiceCount() {
```

File: omod/src/main/java/org/openmrs/module/appointments/web/contract/DailyAppointmentServiceSummary.java [hunk 17]
```diff
@@ -9,13 +9,23 @@ public class DailyAppointmentServiceSummary {
...
         this.appointmentDate = appointmentDate;
         this.appointmentServiceUuid = appointmentServiceUuid;
     }
 
+    public DailyAppointmentServiceSummary(Date appointmentDate, String appointmentServiceUuid, Integer allAppointmentsCount, Integer missedAppointmentsCount, String appointmentServiceName) {
+        this.allAppointmentsCount = allAppointmentsCount;
+        this.missedAppointmentsCount = missedAppointmentsCount;
+        this.appointmentDate = appointmentDate;
+        this.appointmentServiceUuid = appointmentServiceUuid;
+        this.appointmentServiceName = appointmentServiceName;
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/appointments/web/contract/DailyAppointmentServiceSummary.java:26]: Why do we need the `appointmentServiceName` if we have the `appointmentServiceUuid`?

File: omod/src/main/java/org/openmrs/module/appointments/web/mapper/AppointmentMapper.java [hunk 18]
```diff
@@ -272,9 +274,24 @@ private Map createPatientMap(Patient p) {
         map.put("name", p.getPersonName().getFullName());
         map.put("uuid", p.getUuid());
         map.put("identifier", p.getPatientIdentifier().getIdentifier());
-        map.put("age", p.getAge());
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/appointments/web/mapper/AppointmentMapper.java:275]: Why were these properties moved?

File: omod/src/main/resources/webModuleApplicationContext.xml [hunk 19]
```diff
@@ -22,6 +22,10 @@
 
 	<!-- Add here beans related to the web context -->
 
-	<context:component-scan base-package="org.openmrs.module.appointments.web" />
+	<bean class="org.openmrs.ui.framework.StandardModuleUiConfiguration">
+		<property name="moduleId" value="appointments"/>
+	</bean>
+
+	<context:component-scan base-package="org.openmrs.module.appointments.web.controller" />
```
Comment (ibacher) [inline omod/src/main/resources/webModuleApplicationContext.xml:29]: There are Spring beans in this module defined in the `org.openmrs.module.appointments.web` package that are not part of the `controller` sub-package.

File: api/pom.xml [hunk 20]
```diff
@@ -104,7 +104,7 @@
 		<dependency>
 			<groupId>org.bahmni.module</groupId>
 			<artifactId>teleconsultation-api</artifactId>
-			<version>1.0.0-SNAPSHOT</version>
+			<version>${teleconsultation.version}</version>
```
[the remark this answers is not in the document]
Comment (ibacher) [reply api/pom.xml:107]: Also, it shouldn't be necessary to specify the version / type or scope for this dependency since those are all defined in the parent `dependencyManagement` (which is inherited).

File: api/src/main/resources/Appointment.hbm.xml (hunk 15 above)
Comment (ojwanganto) [reply api/src/main/resources/Appointment.hbm.xml:53]: Thanks Ian. Our intention was to have a sure way of letting us know those who come early for their appointments. We probably can keep track of visits and mark appointments as fulfilled if a patient visits just before the appointment date but again that is never 100%. We added a section in the check-in form that lists all future appointments for a patient with a checkbox to mark if any of the future appointments is being fulfilled by the visit. This way, it is easy to report those who come early without necessarily doing a lot of guesses. How else would you approach this?

File: omod/src/main/resources/webModuleApplicationContext.xml (hunk 14 above)
Comment (ojwanganto) [reply omod/src/main/resources/webModuleApplicationContext.xml:25]: We actually should remove this.

File: omod/src/main/resources/config.xml (hunk 13 above)
Comment (ojwanganto) [reply omod/src/main/resources/config.xml:63]: Makes sense

File: api/src/main/java/org/openmrs/module/appointments/dao/impl/AppointmentDaoImpl.java [hunk 21]
```diff
@@ -45,6 +56,61 @@ public List<Appointment> getAllAppointments(Date forDate) {
...
 
+    @Override
+    public List<Appointment> getAllAppointments(Date forDate, String status) {
+        Criteria criteria = sessionFactory.getCurrentSession().createCriteria(Appointment.class);
+        criteria.add(Restrictions.eq("voided", false));
+        criteria.createAlias("patient", "patient");
+        criteria.add(Restrictions.eq("patient.voided", false));
+        criteria.add(Restrictions.eq("patient.personVoided", false));
+        if (forDate != null) {
+            Date maxDate = new Date(forDate.getTime() + TimeUnit.DAYS.toMillis(1));
```
[the remark this answers is not in the document]
Comment (ojwanganto) [reply api/src/main/java/org/openmrs/module/appointments/dao/impl/AppointmentDaoImpl.java:67]: We are working on the updates. Thanks for the review

**Comment by mseaton** [comment]
I'm closing this PR as it is being redone by @ojwanganto and @CynthiaKamau from the correct fork/branch in a series of separate, focused PRs.
